File: js/src/endpoints/vipps.ts
import { IdkollenClient, IdkollenError, PollOptions } from "@/client";
import type { VippsAuthRequest, VippsBackchannelAuthRequest, VippsStatus } from "@/models/vipps";

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

/** Vipps MobilePay operations. */
export class VippsEndpoint {
  constructor(private readonly client: IdkollenClient) {}

  /** Start a Vipps authentication session. */
  public async auth(req: VippsAuthRequest): Promise<VippsStatus> {
    return this.client._post("/v3/vipps/auth", req);
  }

  /** Start a Vipps backchannel authentication session. */
  public async backchannelAuth(req: VippsBackchannelAuthRequest): Promise<VippsStatus> {
    return this.client._post("/v3/vipps/backchannel/auth", req);
  }

  /** Poll the current status of a Vipps authentication session. */
  public async authStatus(id: string): Promise<VippsStatus> {
    return this.client._get(`/v3/vipps/auth/${id}`);
  }

  /** Cancel a Vipps authentication session. */
  public async cancelAuth(id: string): Promise<void> {
    return this.client._delete(`/v3/vipps/auth/${id}`);
  }

  /**
   * Poll until the authentication session reaches a terminal state or the timeout elapses.
   * @throws {IdkollenError} with code `"poll_timeout"` if the timeout is exceeded.
   */
  public async waitForAuth(
    id: string,
    opts: PollOptions = new PollOptions(),
  ): Promise<VippsStatus> {
    const deadline = Date.now() + opts.timeoutMs;

    while (true) {
      const status = await this.authStatus(id);

      if (status.status !== "PENDING") {
        return status;
      }

      if (Date.now() >= deadline) {
        throw new IdkollenError("poll_timeout", 0, "Poll timed out");
      }

      await sleep(opts.intervalMs);
    }
  }
}

File: rust/src/endpoints/vipps.rs
use crate::client::{IdkollenError, PollOptions, WaitError};
use crate::models::{VippsAuthRequest, VippsBackchannelAuthRequest, VippsStatus};
use std::time::Instant;

#[cfg(feature = "async")]
use crate::client::IdkollenClient;

#[cfg(feature = "async")]
/// Vipps operations.
pub struct VippsEndpoint<'a>(pub(crate) &'a IdkollenClient);

#[cfg(feature = "async")]
impl VippsEndpoint<'_> {
    /// Start a Vipps authentication session.
    pub async fn auth(&self, req: VippsAuthRequest) -> Result<VippsStatus, IdkollenError> {
        self.0.post("/v3/vipps/auth", &req).await
    }

    /// Start a Vipps backchannel authentication session.
    pub async fn backchannel_auth(
        &self,
        req: VippsBackchannelAuthRequest,
    ) -> Result<VippsStatus, IdkollenError> {
        self.0.post("/v3/vipps/backchannel/auth", &req).await
    }

    /// Poll the current status of a Vipps authentication session.
    pub async fn auth_status(&self, id: &str) -> Result<VippsStatus, IdkollenError> {
        self.0.get(&format!("/v3/vipps/auth/{}", id)).await
    }

    /// Cancel a Vipps authentication session.
    pub async fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/vipps/auth/{}", id)).await
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub async fn wait_for_auth(
        &self,
        id: &str,
        opts: PollOptions,
    ) -> Result<VippsStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id).await?;

            match status {
                VippsStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    tokio::time::sleep(opts.interval).await;
                },
                terminal => return Ok(terminal),
            }
        }
    }
}

#[cfg(feature = "blocking")]
use crate::client::IdkollenBlockingClient;

#[cfg(feature = "blocking")]
/// Vipps operations (blocking).
pub struct VippsBlockingEndpoint<'a>(pub(crate) &'a IdkollenBlockingClient);

#[cfg(feature = "blocking")]
impl VippsBlockingEndpoint<'_> {
    /// Start a Vipps authentication session.
    pub fn auth(&self, req: VippsAuthRequest) -> Result<VippsStatus, IdkollenError> {
        self.0.post("/v3/vipps/auth", &req)
    }

    /// Start a Vipps backchannel authentication session.
    pub fn backchannel_auth(
        &self,
        req: VippsBackchannelAuthRequest,
    ) -> Result<VippsStatus, IdkollenError> {
        self.0.post("/v3/vipps/backchannel/auth", &req)
    }

    /// Poll the current status of a Vipps authentication session.
    pub fn auth_status(&self, id: &str) -> Result<VippsStatus, IdkollenError> {
        self.0.get(&format!("/v3/vipps/auth/{}", id))
    }

    /// Cancel a Vipps authentication session.
    pub fn cancel_auth(&self, id: &str) -> Result<(), IdkollenError> {
        self.0.delete(&format!("/v3/vipps/auth/{}", id))
    }

    /// Poll until the authentication session reaches a terminal state or the timeout elapses.
    pub fn wait_for_auth(&self, id: &str, opts: PollOptions) -> Result<VippsStatus, WaitError> {
        let deadline = Instant::now() + opts.timeout;

        loop {
            let status = self.auth_status(id)?;

            match status {
                VippsStatus::Pending(_) => {
                    if Instant::now() >= deadline {
                        return Err(WaitError::Timeout);
                    }

                    std::thread::sleep(opts.interval);
                },
                terminal => return Ok(terminal),
            }
        }
    }
}
